### DISCUSSION TITLE
What is the general process for dealing with node reliant / typical backend processes? (Trying to get sqlite to work)

### DISCUSSION THREAD

**Discussion by rmcdonald1442** [socketsupply/socket#889] [Q&A]
1)

The guide guides on socket-supply.co/guides recomments SQL-Lite or level db for having some level of persistent storage. The standard sqlite3 npm uses node's process variable. Because this is not available a number of errors exist when running in the browser.

Is there some simple approach to make process available (and populated with (e.g) environment variables) for a frontend app? Or should I be creating a server that runs in the socket runtime for anything that requires node/process/whatever else?

2)

I've been trying to achieve the latter (runtime server) approach - initially just making a basic "Hello world" server which by frontend can ping. Unfortunately the `socket.ini` "cmd" setting does not appear to be playing nicely - I have tried the following but have seen no success, even with the --verbose flag I can't tell if this command is ever running. Am I doing something wrong?:

```
[win]

; The command to execute to spawn the “back-end” process.
; cmd = "node server.js"
; cmd = "node src/backend/server.js"
; cmd = "node ./src/backend/server.js"
cmd = "echo SOMETHING"
```

This is with `server.js` being copied into dist, which eventually becomes build directory (I can verify this witha screenshot): 

```
import { defineConfig } from "vite";
import react from "@vitejs/plugin-react";
import externalize from "vite-plugin-externalize-dependencies";
import copy from "rollup-plugin-copy";

export default defineConfig({
  define: {
    process: {},
  },
  plugins: [
    externalize({
      externals: [/^socket:.*/],
    }),
    react(),
    copy({
      targets: [{ src: "./src/backend/**", dest: "./dist" }],
      hook: "writeBundle",
    }),
  ],
  build: {
    outDir: "./dist",
    rollupOptions: {
      external: [/^socket:.*/],
    },
  },
});
```

**Comment by rmcdonald1442** [comment]
I've seen this now: https://socketsupply.co/guides/#

**Comment by mehdi-cit** [comment]
Hi @rorymcd98 
Could you please be explicit on how you were able to make SQL-Lite work "within" socket runtime. I'd be very much appreciative if someone could layout how to exactly use SQL-Lite. It's best if one can use the SQL-Lite from the browser and not the "backend/server"!
Specifically, I am trying to see how to create an android application with a socketsupply that uses SQL-Lite!

**Reply by rmcdonald1442** [reply]
Hi,

My last post here was not very clear - I was using a backend approach.

However, I have since began to move on from this approach and towards [sql.js](https://github.com/sql-js/sql.js), which allows for interacting with a sqlite db from a browser via a static wasm file. I'm sure if you were very clever you could integrate this into the socket runtime somehow, but I just load it inside my `.js` files.  You can save / load by using sql.js's `db.export()` and saving the outputted binary to storage with `socket:fs`. 

I was able to import this with Vite by doing something like this (from memory, there may be mistakes): 

`npm install sql.js`

Then copy ./node_modules/sql.js/dist/sql-wasm.wasm to my `./src` folder

```
import wasmUrl from "./sql-wasm.wasm?url"
const SQL = await initSqlJs({
  locateFile: () => wasmUrl
});
```


Lastly, update your content security policy in HTML (can't remember how I did this but I will post next time I work on socket-supply stuff).

@chicoxyzzy

**Comment by mehdi-cit** [comment]
Thank you for the reply @rorymcd98 However this is not at all safe! At least not for my use case.
In case you are not aware, any disk space used from within the browser might be reclaimed from the OS at anytime. This in practice does happen when the -say- Android mobile reaches a high percentage of disk usage. In contrast, the OS won't destroy/erase data managed by "native" apps. Well, the user would need to manually allow the mobile to clear the data in case of native apps. 

Could someone from socketsupply team chime-in and clarify how can one use sqlite as a relational database from within the "socket app"!


PS: Sorry just saw the code snippet using socket:fs ! Well in such case you are using it from within a "native app" and it should be fine. But this feels hakish to me and I am sure the socketsupply can do better ❔
